Code/Framework/AzCore/Tests/AZStd/Hashed.cpp

```diff
@@ -287,6 +287,54 @@ namespace UnitTest
         }
     }
 
+    TEST_F(HashedContainers, HashTable_InsertionDuplicateOnRehash)
+    {
+        struct TwoPtrs
+        {
+            void* m_ptr1;
+            void* m_ptr2;
+
+            bool operator==(const TwoPtrs& other) const
+            {
+                if (m_ptr1 == other.m_ptr1)
+                {
+                    return m_ptr2 == other.m_ptr2;
+                }
+                else if (m_ptr1 == other.m_ptr2)
+                {
+                    return m_ptr2 == other.m_ptr1;
+                }
+                return false;
+            }
+        };
+
+        // This hashing function produces different hashes for two equal values,
+        // which violates the requirement for hashing functions.
+        // The test makes sure that this does not reproduce an issue that caused the insert() function to loop infinitely.
+        struct TwoPtrsHasher
+        {
+            size_t operator()(const TwoPtrs& p) const
+            {
+                size_t hash{ 0 };
+                AZStd::hash_combine(hash, p.m_ptr1, p.m_ptr2);
+                return hash;
+            }
+        };
+        using PairSet = AZStd::unordered_set<TwoPtrs, TwoPtrsHasher>;
+        PairSet set;
+        set.insert({ (void*)1, (void*)2 });
+        set.insert({ (void*)3, (void*)4 });
+        set.insert({ (void*)5, (void*)6 });
+        set.insert({ (void*)7, (void*)8 });
+        // Elements with different hashes, but equal
+        set.insert({ (void*)0x000001ceddd9ca20, (void*)0x000001ceddd9cba0 }); // hash(148335135725641)
+        set.insert({ (void*)0x000001ceddd9cba0, (void*)0x000001ceddd9ca20 }); // hash(148335135764189)
+        AZ_TEST_START_TRACE_SUPPRESSION;
+        // This will trigger the assertion of duplicated elements found
+        set.rehash(23);
```

Comment:
Nice thanks for updating this :+1:

Reply:
Can you provide a comment on why 23 buckets were supplied to rehash as opposed to any other number?
Is this the number of buckets needed to repro this issue? Is it Random?

Reply:
That bucket size triggers rehashing of the elements, rehash() function will only actually rehash if the number is in the new bracket of possible sizes: (atm its these prime numbers: 7, 23, 53, 97, 193, 389, 769,...).

23 is the case of having a hash collision between the different hashes of the equal elements